src/osd/ReplicatedPG.cc

```diff
@@ -1357,6 +1357,35 @@ int ReplicatedPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
   return r;
 }
 
+// it is pretty much the same as CEPH_OSD_OP_COPY_FROM, but it requires user to
+// specify the src osd, and uses repair_read() to read the object payload, also
+// the src oid is always identical to the dest oid (the one in OpContext)
+int ReplicatedPG::do_repair_copy(OpContext *ctx, const OSDOp& osd_op, bufferlist::iterator& bp)
+{
+  uint32_t what;
+  vector<pg_shard_t> bad_shards;
+  try {
+    ::decode(what, bp);
```

Comment:
maybe i should wrap it in an arg struct, and add the `encode()/decode()` methods for it.